🎉 Source S3: Add IAM Instance Profile Support

sidartha opened the pull request:
## What
This PR adds IAM instance profiles support for the S3 source connector to allow use of the instance profile instead of relying on Access Keys and Secrets (as described in #5282 and #5942). A similar feature was implemented for destination-s3 in #9399. 
 
<img width="819" alt="Screenshot 2022-02-27 at 10 22 01 AM" src="https://user-images.githubusercontent.com/12951761/155865684-c8497f49-528e-4c46-ba43-07dade9441ff.png">

## How

A new config option `use_aws_default_credential_provider_chain` is added and when set to true, we rely on Boto3 to identify the credentials (this is done by removing `signature_version=UNSIGNED` in the client config object). 

The other option considered was to fall-back to the default credential provider chain if no access key & secret are provided but this would cause issues for those who are attempting to read a publicly available S3 file without using any credentials and also for those who are using this connector with non-AWS S3 storage.

I'm open to other ideas for implementing this too - this was just the simplest thing I could think of!

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)



## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=10676) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/airbytehq/airbyte?pullRequest=10676) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/airbytehq/airbyte?pullRequest=10676) it.</sub>

ntkawasaki commented:
Hey @sidartha I think the code changes here are straightforward and solid 👍  and thank you for adding a test

One thought I have is to change the `boolean` type of the connection option `use_aws_default_credential_provider_chain` to become a [`oneOf`](https://docs.airbyte.com/connector-development/connector-specification-reference#using-oneofs) option like `authentication_style` where the users would choose `use_aws_default_credential_provider_chain` as the authentication strategy (maybe set this as the default, also). This would set a nice base for adding new future auth styles to the source very easily!

Also, you'll need to update a couple more places around the repo such as the Dockerfile image tags, a STANDARD_SOURCE_DEFINTION.yaml file, and the spec.json file in the connector directory. See this PR for an example of the other places to update: https://github.com/airbytehq/airbyte/pull/9623/files#diff-0a23a52edd477f81527cdead182adda4e4d84ec23e10139bdbfecfd33ef6f071L5 (Note that this PR updates a destination and a source together, you only need to worry about the source files)

alafanechere commented on airbyte-integrations/connectors/source-s3/source_s3/s3file.py:
```diff
@@ -33,6 +33,9 @@ def _setup_boto_session(self) -> None:
                 aws_secret_access_key=self._provider.get("aws_secret_access_key"),
             )
             self._boto_s3_resource = make_s3_resource(self._provider, session=self._boto_session)
+        elif self.use_aws_default_credential_provider_chain:
+            self._boto_session = boto3session.Session()
+            self._boto_s3_resource = make_s3_resource(self._provider, config=Config(), session=self._boto_session)
         else:
             self._boto_session = boto3session.Session()
             self._boto_s3_resource = make_s3_resource(self._provider, config=Config(signature_version=UNSIGNED), session=self._boto_session)
```
I don't get the difference between the `else` and the `elif` block.

alafanechere commented on airbyte-integrations/connectors/source-s3/source_s3/s3file.py:
```diff
@@ -43,6 +46,10 @@ def use_aws_account(provider: Mapping[str, str]) -> bool:
         return True if (aws_access_key_id is not None and aws_secret_access_key is not None) else False
 
+    @staticmethod
+    def use_aws_default_credential_provider_chain(provider: Mapping[str, str]) -> bool:
```
What do you think of refactoring this a bit: 
* create a `AuthenticationMethod` enum 
* create a method that returns the authentication method to use (`get_authentication_method`)
* use `get_authentication_method` in `_setup_boto_session` to create session according to the authentication method to use

eg: 
```python 
class AuthenticationMethod(Enum):
  ACCESS_KEY_SECRET_ACCESS_KEY = 1
  DEFAULT = 2
  
class S3File:
  def __init__(self, *args: Any, **kwargs: Any):
    super().__init__(*args, **kwargs)
    self.authentication_method = self.get_authentication_method()
    self._setup_boto_session()

  def get_authentication_method(self):
    if self._provider.get("aws_access_key_id") and self._provider.get("aws_secret_access_key"):
      return AuthenticationMethod.ACCESS_KEY_SECRET_ACCESS_KEY
    elif self._provider.get("use_aws_default_credential_provider_chain"):
      return AuthenticationMethod.DEFAULT
    else:
      raise Exception("Could not determine an authentication method according to current provider.")

  def _setup_boto_session(self):
    if self.authentication_method == AuthenticationMethod.ACCESS_KEY_SECRET_ACCESS_KEY:
      ...
    if self.authentication_method == AuthenticationMethod.DEFAULT:
      ...
```

alafanechere replied:
I'd be great if we could define at a single spot the specifics related to an authentication method (sessions and params definition) to make the rest of the implementation independent from the authentication method used. My point is to avoid several if / else statements on `use_aws_account` or `use_aws_default_credential_provider_chain`. It will improve maintainability  and eventual iterations on authentication methods.

alafanechere reviewed (changes requested):
Thank you @sidartha for this contribution! 
I totally agree with @ntkawasaki about oneOf field to pick authentication method from.
I made a refactoring suggestion, it'd be great if you could find a way of not declaring multiple if/else statements on the authentication method.

I think we shall also add more acceptance test using a config that have `use_aws_default_credential_provider_chain = true`.

Finally, could you please bump the connector version (in the Dockerfile and in `airbyte-config/init/src/main/resources/seed/source_definitions.yaml` ), and update it's documentation (`docs/integrations/sources/s3.md`) . Please explain what is the default chain of credentials that Boto will try to use, and update the changelog.

Let me know what you think of my suggestions!

sidartha commented:
Thanks for the feedback @ntkawasaki and @alafanechere! I have updated the PR based on your suggestions.

Since both `s3file.py` and `stream.py` require auth to access S3, I moved the auth related logic to `s3_utils.py`.  

For the acceptance test, do you have any suggestions on how to add a `secrets/config.json` file with `use_aws_default_credential_provider_chain = true`?

alafanechere commented on airbyte-integrations/connectors/source-s3/integration_tests/spec.json:
```diff
@@ -177,6 +177,11 @@
             "default": "",
             "type": "string"
           },
+          "use_aws_default_credential_provider_chain": {
```
Could you please create an `authentication_method` oneOf field which for the user to chose between *Default credential provider* and *Access Key/Secret Key*?

sidartha replied:
@alafanechere I am a little stuck here on how to implement this without causing a breaking change. 

We have Default Credential Provider, Access Key/Secret Key and no authentication. The current behaviour defaults to Access Key/Secret Key if those values are included and falls back to no authentication if not included. 
When introducing oneOf field `authentication_method`, I'm not sure what the default value can be set to.

alafanechere replied:
What do you think of removing `use_aws_default_credential_provider_chain` field and changing the description of Access Key/Secret Key to explain that if not set it will fall back to the default boto credential chain?

sidartha replied:
Since we need to support those who don't use any credentials (e.g. those who use it with non-AWS storage or those who don't need any auth since they are accessing public buckets), I don't think this will work. In this case, the default boto credential chain won't be able to locate any credentials.

alafanechere reviewed (changes requested):
Thank @sidartha for the change! I've still one remaining blocking change in the `spec.json`.
I'll try to run the acceptance test with our current config using access key/secret key. If they pass I'll try to add a new one with `use_aws_default_credential_provider_chain`.

alafanechere commented:
/test connector=connectors/source-s3

alafanechere commented:
/test connector=connectors/source-s3

alafanechere commented:
/test connector=connectors/source-s3

alafanechere commented:
/test connector=connectors/source-s3

alafanechere commented:
/test connector=connectors/source-s3

alafanechere commented:
Hey @sidartha I run the acceptance tests to make sure your PR does not introduce any breaking change. 
I'd still like the `spec.json` to be refactored a bit to introduce an Authentication Method dropdown menu with `oneOf` fields. I'll try to make the change myself if you don't feel like doing it.
I also still need to figure out a way to run the acceptance test with a config that is using the default credentials provider you introduced. This will probably lead me to some specific IAM settings I need to declare for the run of the tests.

sidartha commented:
Hi @alafanechere, I'd appreciate the help on editing the `spec.json` to add a `oneOf` authentication method field (or do you have any pointers to other sources that have implemented this where we are overriding the `SourceFilesAbstractSpec`?). Thanks!

matthewhembree commented:
**Question:**
I don't intend to expand the scope of this PR, but I am interested if post-merge refactoring could be avoided.

What issues do you see could be encountered if an additional method like `AuthenticationMethod.ACCESS_KEY_SECRET_ACCESS_KEY_ASSUME_ROLE` would be added? I'd like to be able to specify in the UI:
- access key id
- secret access key
- role to assume (ARN)

And those credentials could establish temporary STS credentials to pass to the S3 client.

My particular use case is cross-account role access.
Source: Account B
Destination: Account A
Airbyte instance homed in Account A

My org has a third party pulling data from our S3 bucket and if they could just assume a role in our account, then we wouldn't have to worry about rotating secret keys.

This is further useful if you have a single Airbyte instance that has sources/destinations in multiple foreign accounts. The default EC2 instance profile can only be tied to a single role ARN in the `~/.aws/config` file which limits the instance to a single foreign account. It also separates the credentials on the instance in two places.

Maybe additional profiles could be added in `~/.aws/config` and reference a `source_profile` of the instance profile.
e.g. This might be a good compromise, but still separates credentials in two places:
```ini
[default]
credential_source = Ec2InstanceMetadata

[profile accountb]
role_arn = arn:aws:iam::111111111111:role/ROLENAME
external_id = bbbbbbbb
source_profile = default

[profile accountc]
role_arn = arn:aws:iam::222222222222:role/ROLENAME
external_id = cccccccc
source_profile = default
```
...and then the UI could take a profile as input.

Would your implementation need to be refactored to add my use case at a future time?

Thanks!

alafanechere commented:
@matthewhembree thank you for this feedback! This is definitely a valid use case and goes in the direction that we shall implement a dropdown field to select authentication methods. I've been really swamped in the past week to suggest the change in `spec.json` but I will soon!

alafanechere commented:
@sidartha to implement the `oneOf` field for the `autentication_method` is suggest you to have a look at how the `format` field is handled on the `SourceFilesAbstractSpec` parent class. You'll have to declare a base models for each authentication method and declare a `authentication_method` field in `SourceS3Spec` whose type is the union of the base model you created. You'll then have to create a custom function similar to `change_format_to_oneOf` in `airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/spec.py` and call it from an overridden `schema` class method.

alafanechere commented:
I'm closing this PR as I don't have the bandwidth to iterate on the specs. @sidartha feel free to re-open if you want to tackle this.

mc51 commented:
Just FYI: I'm working on finishing this up. I'll open a new PR soon.
